Examplary (non-Evidence) interaction model vor CoSERV (#91)
ietf-rats-wg/draft-ietf-rats-reference-interaction-models: draft-ietf-rats-reference-interaction-models.md

    tpm2-signature: bytes,
    ? ak-cert: bytes,         ; TPM2 attestation key certificate (AK Cert)
]

# An Examplary Interaction Model for Endorsement Retrieval with Implicit Handles

The interaction models specified in this document focus on Evidence conveyance.
Evidence conveyance requires an indication of recentness, which a Verifier can use to determine the freshness of Evidence during its appraisal.
This document introduces the information element `handle` to include this extra-data in the conveyance of evidence.
However, the conveyance of other Conceptual Messages (see {{-rats-cmw}}) does not necessarily rely on the concept of freshness to the same extent as Evidence.
Examples include: Endorsements and Reference Values.

CoSERV ({{-CoSERV}}) defines a request/response model for conveying, caching, and potentially aggregating endorsements and reference values, as well as a retrieval and query language for these values that works across various Conceptual Message providers and RATS owner roles.

The absence of an explicit `handle` in the CoSERV query language does not imply a lack of corresponding semantics.

> The problem with including an explicit `handle` information element in CoSERV is that it would undermine CoSERV's caching mechanism.

Reviewer: > The problem with including an explicit `handle` information element in CoSERV is that it would undermine CoSERV's caching mechanism.

Can we spell out _how_ it is undermined, just to be kinder to readers? In the CoSERV draft we have a specific section §3.4.3 on "Avoidance of Volatile Data in Queries". It might help to reference that specific section from here, just to help glue things together.

Reviewer 2: fixed in 99a2a7eb1dc0cebf162914d8869b5a0730aa5c6c

> Instead, the combination of the `attEnvIDs` and `claimSelection` information elements, which are part of the CoSERV request, can be interpreted as representing the start of an epoch (see {{-epoch-markers}}).

Reviewer: We could do with spelling this out, because I had to scratch my head a bit to understand how `claimSelection` is part of a CoSERV request. CoSERV supports queries by "stateful environments", so I'm assuming that `claimSelection` corresponds to the stateful part. But it would reassure me (and help other readers) to say this explicitly. We could enrich this further by referencing §3.3.1 of CoSERV ("Stateful Environments").

Reviewer 2: Addressed in 5675f277776b9feaeed9fd75c700e21c635636b4

Essentially, the URL in the GET request becomes the `handle` in the interaction.
The `generateEndorsements` operation illustrated in {{fig-coserv-transaction}} signs a set of Endorsements and Reference Values (typically, one or more CoRIM tags {{-CoRIM}}) alongside the query and expiration timestamp for the response.
Semantically, this makes the URL in the GET request a `handle` that specifies the start of an epoch, while the expiration timestamp specifies the end.
This approach assumes that an entity taking on a RATS role has access to a reliable time source, for example, using an Epoch Bell {{-epoch-markers}}.

~~~~ aasvg
.----------.                                                .----------.

> |  CoSERV  |                                                | Verifier |

Reviewer: Slight inconsistency of labelling here in that `Verifier` is a RATS role while `CoSERV` is not. CoSERV is the conveyance protocol, so it feels awkward to use it as the label for only one of the participants. It operates between the participants.

We could label them generically as `CoSERV producer` and `CoSERV consumer`, with some prose to explain that the producer is typically an `Endorser` or `Reference Value Provider`, while the consumer is typically a `Verifier`.

Or, we could make the labels much more wordy and have them as `CoSERV producer (e.g. Endorser or Reference Value Provider)`, and `CoSERV consumer (e.g. Verifier)`.

Reviewer 2: addressed in 8133f7a3672df8768d231a506d930c0dddc69d21

'----+-----'                                                '-----+----'
     |                                                            |
====================[Initiating Evidence Appraisal]=====================
     |                                                            |
     |                                  initEvidenceAppraisal(evidence)
     |                               attEnvIDs, claimSelection <= |
     |                                                            |
===============[Endorsements Collection and Conveyance]=================
     |                                                            |

>   ingestEndorsements(attEnvIDs, claims)                           |

Reviewer: I'd contend that `ingestEndorsements` is a separate phase of its own (e.g. `========[Endorsements Ingestion]======`). With the diagram as structured right now, it implies (to my eye) that there is an ingestion event per each request/collect event, which is not the case.

Reviewer 2: Addressed in bc964272ab5d82cca89c27a19f38b225846f5840

     | => endorsements                                            |
     |                                                            |

>      |<----------------- requestEndorsements(attEnvIDs, claimSelection)

Reviewer: As of version 06 of CoSERV, we now also support queries by RIM identifiers instead of by stateful environments. Should we be representing this in the diagram? (I think this diagram was originally created before such queries existed).

The issue is that CoSERV queries are now bifurcated, with environment-based queries and RIM-ID-based queries as two independent and exclusive forks of the data model. It's effectively a union type. I'm not sure if there's a convention for representing that here. Feels like we want something like:

`requestEndorsements((attEnvIDs, claimSelection) | (rimIDs))`

Or maybe there's a better canonical term for `rimIDs` that we should re-use here?

Reviewer 2: addressed in 05c50929f8e3af72bd5e5046f1374978c6eabf06

    tpm2-signature: bytes,
    ? ak-cert: bytes,         ; TPM2 attestation key certificate (AK Cert)
]

# An Examplary Interaction Model for Endorsement Retrieval with Implicit Handles

The interaction models specified in this document focus on Evidence conveyance.

> Evidence conveyance requires an indication of recentness, from which freshness if Evidence can be deducted by a Verifier during its appraisal of Evidence.

Reviewer 2: 1st attempt at de-clunkification

```suggestion
Evidence conveyance requires an indication of recentness, which a Verifier can use to determine the freshness of Evidence during its appraisal.
```

But I am not even sure this is what you really meant.

Author: this is what I meant... it was poorly phrased.

> In this document, the information element `handle` is introduced to include such extra-data in the conveyance of Evidence.

Reviewer 2: use the active form

```suggestion
This document introduces the information element `handle` to include this extra-data in the conveyance of evidence.
```

Author: :+1:

The conveyance of other Conceptual Messages (see {{-rats-cmw}}) does not necessary rely on the concept of freshness as much as Evidence.

> Examples include: Endorsements and Reference Values

Reviewer 2: ```suggestion
However, the conveyance of other Conceptual Messages (see {{-rats-cmw}}) does not necessarily rely on the concept of freshness to the same extent as Evidence.
Examples include: Endorsements and Reference Values.
```

The Concise Selector for Endorsements and Reference Values ({{-CoSERV}}) document specifies a structured request/response model for not only the conveyance of Endorsements and Reference Values, but also for their in-path caching and potentially aggregation.

> As such, the structured request/response model defined by CoSERV effectively defines a retrieval and query language for Endorsements and Reference Values that are distributed across various Concptual Message providers and RATS owner roles.

Reviewer 2: streamline

```suggestion
CoSERV ({{-CoSERV}}) defines a request/response model for conveying, caching, and potentially aggregating endorsements and reference values, as well as a retrieval and query language for these values that works across various Conceptual Message providers and RATS owner roles.
```

Author: better

While there is no explicit 'handle' included in the CoSERV query language, that does not imply that the corresponding semantics are missing from the language.

> Explicitly including a 'handle' information element in CoSERV would undermine an optimization scheme that CoSERV realizes: the combination of the information elements 'attEnvIDs' and 'claimSelection', which are part of the CoSERV request, represent the start of an Epoch (see {{-epoch-markers}}.

Reviewer 2: de-clunkify

```suggestion
The absence of an explicit `handle` in the CoSERV query language does not imply a lack of corresponding semantics.
The problem with including an explicit `handle` information element in CoSERV is that it would undermine CoSERV's caching mechanism.
Instead, the combination of the `attEnvIDs` and `claimSelection` information elements, which are part of the CoSERV request, can be interpreted as representing the start of an epoch (see {{-epoch-markers}}).
```

In essence, the URL in the get request becomes the 'handle' in the interaction.
CoSERV's handle variant (effectively, the URL in the GET request) is used to minimize in-path interaction via caching.
The 'generateEndorsements' operation illustrated below {{fig-coserv-transactionfig-coserv-transaction}} signs a set of Endorsements and Reference Values (typically, one or more CoRIM tags {{-CoRIM}}) alongside the query and and expiration timestamp for the response.
Semantically, this renders the URL in the GET request effectively a 'handle' that specifies the start of an epoch and the expiration timestamp specifies the end of an epoch.

> Of course, this approach implies that an entity taking on a RATS role has access to a reliable time source (which could be an Epoch Bell {{-epoch-markers}}.

Reviewer 2: ```suggestion
Essentially, the URL in the GET request becomes the `handle` in the interaction.
The `generateEndorsements` operation illustrated in {{fig-coserv-transaction}} signs a set of Endorsements and Reference Values (typically, one or more CoRIM tags {{-CoRIM}}) alongside the query and expiration timestamp for the response.
Semantically, this makes the URL in the GET request a `handle` that specifies the start of an epoch, while the expiration timestamp specifies the end.
This approach assumes that an entity taking on a RATS role has access to a reliable time source, for example, using an Epoch Bell {{-epoch-markers}}.
```
